### ISSUE TITLE
v2.0.0 Roadmap

### ISSUE THREAD

**Issue by Yomguithereal** [Yomguithereal/baobab#262]
* State/facets convergence (#240).
* Dropping `$cursor`.
* Better code self documentation.
* Dropping `$chain`.
* Probably converge towards `syncwrite`.
* `immutable` by default?

**Comment by ivan-kleshnin** [comment]
> immutable by default?

:+1: 

I also want to propose to add a list of projects using Baobab to the main README.
It's important to see there is some community around project when you make a library choice.

**Comment by Yomguithereal** [comment]
Why not @ivan-kleshnin. However I must admit I am not well informed of the projects/persons using Boabab.

**Comment by ivan-kleshnin** [comment]
Let's ask people then @tnrich @ostinru @lewisl9029 @jeroencoumans @kevbook @sheerun @vjpr @christianalfoni  @Aetet @jacomyal @chasm @gersongoulart @luniki @nthtran @RJ @milanlandaverde @serjout @bebraw @JohnAmican @asfktz @AutoSponge @alexmingoia @cassiozen @DouglasLivingstone @scabbiaza anyone wants to share his or her Baobab-powered project?

I can propose this one https://github.com/Paqmind/react-ultimate which is React + Baobab + ES6
combination.

**Comment by Aetet** [comment]
@ivan-kleshnin 
:+1:  for baobab-powered project.

**Comment by chasm** [comment]
Yes, please. Also, yes to immutable by default.

**Comment by doobleweb** [comment]
yes to immutable by default

**Comment by sheerun** [comment]
Yup. There's not much benefit in mutability. Any reason to keep it?

**Comment by Yomguithereal** [comment]
The reason to keep it is performance. Immutability comes with some performance loss. The ideal workflow is to dev with an immutable tree and let the prod be mutable so you can have better performance.

**Comment by Aetet** [comment]
I really don't want immutability by default. May be only for development environment. And for production us mutable version. Like React and their warnings. 
In my code I haven't much places, that I break rule for immutability.

**Comment by ivan-kleshnin** [comment]
In any case we'd use 

```
new Baobab(..., {
  immutable: NODE_ENV != production
})
```

so it does not matter what default is :wink:

**Comment by Aetet** [comment]
:+1:

**Comment by AutoSponge** [comment]
My $0.02... I used Baobab cursors to hook small components together on their own render cycles. Adding immutability to that adds no performance since I can filter the stream to only emit differences. This makes Baobab + virtual-dom as fast as native web components without doing anything else.

Adding immutability *could* allow for rollback/transactional control of the model. However, this would have to be abstracted inside Baobab so I don't need to pass around new cursor references to every component. Since that is only a feature and not a performance improvement, it should be optional IMO.

**Comment by RJ** [comment]
my reckons for 2.0:
* Immutable by default (mutability is for performance later, if really needed)
* Syncwrite by default - anything else is confusing
* fine with dropping $cursor

i'm not up-to-date on the other points, so no opinion on them yet (need to read #240 properly)

**Comment by Aetet** [comment]
And I think facets must have opportunity to return function for passing user arguments. It's very handy when we have nested structure and nested field can have computed values. like this:
```js
var tree = new Baobab({
  cells: [{
     cards: [{id: 1, title: 'first'}, {id: 2, title: 'second'}],
     colCount: 1
     // rowCount must be here, but it is computable. 
     // Based on cards.length and colCount
  }]
});
```
```js
rowCount: {
cursors: {
   cells: ['cells']
},
get: function(cells) {
   return function (id) {
     let cell = cells[id];
     let colCount = cell.colCount
     return cell.cards.length / colCount;
  }
}
}
```
And in widget we can do something like:
```js
class A extends React.component{
  render() {
     let {cells} = this.props;
     let cellViews = cells.map(cell => <Cell rowCount={rowCount(cell.id)} />)
  }
}
branch(A, {
  cursors: {
     cells: ['cells']
  },
  facets: {
     rowCount: 'rowCount'
  }
  
})
```

So what's the best way to do that?

**Comment by sporto** [comment]
+1 immutable by default

**Comment by Yomguithereal** [comment]
I am currently writing `v2` and wondering whether I should keep the `relevant` and `irrelevant` events. Does someone actually use them?

**Comment by ivan-kleshnin** [comment]
> Does someone actually use them?

Me not.

**Comment by AutoSponge** [comment]
I'm using them in my current app. I've found there are certain scenarios where the expected 'update' event does not fire but the 'irrelevant' event does fire. So far, in all cases, the 'relevant' event fires just before the update event so it's not as useful.

**Comment by Yomguithereal** [comment]
Are you saying that the `update` event does not behave correctly and that you need the `relevant/irrelevant` events to actually get things working @AutoSponge? My question here is, should I fix something about the `update` event so we can get rid of the other events which are only used as a workaround?

**Comment by Yomguithereal** [comment]
By the way @AutoSponge, with you agreement, I will probably introduce some of the concepts of `perfget` in some of the v2's use cases.

**Comment by AutoSponge** [comment]
@Yomguithereal re: events: Yes, I noticed that `unset` does not fire `updated`. Here's a fiddle to explain: http://jsfiddle.net/ube1k2pr/ (but it's not loading for me). It doesn't matter if you `unset` from the cursor or the tree. I think if this worked as expected, I would not need `irrelevant`.

```javascript
var tree = new Baobab();
var key = 'a.b.c'.split('.');
var $abc = tree.select(key);
var log = document.write.bind(document, '<hr/>');
$abc.on('update', function (e) {
    log('updated:', e.data.data);
});
tree.set({a:{b:{c:1}}});
setTimeout(function () {
    $abc.unset();
    //tree.unset(key);
    setTimeout(function () {
        log('final: ', $abc.get());
    }, 100);
}, 100);
```

@Yomguithereal I'd be happy to help do some benchmarking for the accessors.

**Comment by Yomguithereal** [comment]
Ok @AutoSponge. I open a specific issue (#274) to deal with the problem. By the way, I quite like your way to indicate that a variable is a cursor through the use of `$`. Kinda remind me of jQuery's styles.

**Comment by AutoSponge** [comment]
I started doing that with streams and I would like a different symbol for cursors but `¢` is harder to type :)

**Comment by Yomguithereal** [comment]
Brace yourselves, my lieges, for beta `v2` is coming very soon.

**Comment by tnrich** [comment]
How soon??!!

**Comment by Yomguithereal** [comment]
Next week.

**Comment by tnrich** [comment]
Sweet!

**Comment by Yomguithereal** [comment]
v2 proposal is now available in #278.
